src/inspector_profiler.cc

```diff
@@ -184,92 +231,28 @@ void V8CoverageConnection::End() {
-  Context::Scope context_scope(context);
-  Local<String> result;
-  if (!String::NewFromTwoByte(isolate,
-                              message.characters16(),
-                              NewStringType::kNormal,
-                              message.length())
-           .ToLocal(&result)) {
-    fprintf(stderr, "Failed to convert profiling message\n");
-  }
-  WriteCpuProfile(result);
+std::string V8CpuProfilerConnection::GetDirectory() const {
+  return env()->cpu_prof_dir();
 }
 
-void V8CpuProfilerConnection::WriteCpuProfile(Local<String> message) {
-  const std::string& filename = env()->cpu_prof_name();
-  const std::string& directory = env()->cpu_prof_dir();
-  CHECK(!filename.empty());
-  CHECK(!directory.empty());
-  uv_fs_t req;
-  int ret = fs::MKDirpSync(nullptr, &req, directory, 0777, nullptr);
-  uv_fs_req_cleanup(&req);
-  if (ret < 0 && ret != UV_EEXIST) {
-    char err_buf[128];
-    uv_err_name_r(ret, err_buf, sizeof(err_buf));
-    fprintf(stderr,
-            "%s: Failed to create cpu profile directory %s\n",
-            err_buf,
-            directory.c_str());
-    return;
-  }
-  MaybeLocal<String> result = GetResult(message);
-  std::string target = directory + kPathSeparator + filename;
-  if (!result.IsEmpty()) {
-    WriteResult(target.c_str(), result.ToLocalChecked());
-  }
+std::string V8CpuProfilerConnection::GetFilename() const {
+  return env()->cpu_prof_name();
 }
 
-MaybeLocal<String> V8CpuProfilerConnection::GetResult(Local<String> message) {
-  Local<Context> context = env()->context();
-  Isolate* isolate = env()->isolate();
-  Local<Value> parsed;
-  if (!v8::JSON::Parse(context, message).ToLocal(&parsed) ||
-      !parsed->IsObject()) {
-    fprintf(stderr, "Failed to parse CPU profile result as JSON object\n");
-    return MaybeLocal<String>();
-  }
-
-  Local<Value> result_v;
-  if (!parsed.As<Object>()
-           ->Get(context, FIXED_ONE_BYTE_STRING(isolate, "result"))
-           .ToLocal(&result_v)) {
-    fprintf(stderr, "Failed to get result from CPU profile message\n");
-    return MaybeLocal<String>();
-  }
-
-  if (!result_v->IsObject()) {
-    fprintf(stderr, "'result' from CPU profile message is not an object\n");
-    return MaybeLocal<String>();
-  }
-
+MaybeLocal<Object> V8CpuProfilerConnection::GetProfile(Local<Object> result) {
   Local<Value> profile_v;
-  if (!result_v.As<Object>()
-           ->Get(context, FIXED_ONE_BYTE_STRING(isolate, "profile"))
+  if (!result
+           ->Get(env()->context(),
+                 FIXED_ONE_BYTE_STRING(env()->isolate(), "profile"))
            .ToLocal(&profile_v)) {
     fprintf(stderr, "'profile' from CPU profile result is undefined\n");
-    return MaybeLocal<String>();
+    return MaybeLocal<Object>();
   }
-
-  Local<String> result_s;
-  if (!v8::JSON::Stringify(context, profile_v).ToLocal(&result_s)) {
-    fprintf(stderr, "Failed to stringify CPU profile result\n");
-    return MaybeLocal<String>();
+  if (!profile_v->IsObject()) {
```

Comment:
CHECK? Should be an impossible condition under normal circumstances, right?

Reply:
It could happen if the protocol changes or if for some reason the response returns an error. I don't think this should crash the process? As it's just the profiler encountering something unexpected, the program should still run normally.

Reply:
A crash gives me a core dump that lets me `bt` and `call profile_v->Print()` :-)

I don't feel too strongly but all other things being equal I'd CHECK.

Reply:
FWIW when I need to debug responses here I usually just modify the source and JSON::Stringify() the whole message. I think it may be worth the cost to always parse the response (so that we can use `id` instead of a `ending_` which relies on the fact that the protocol implementation is synchronous) and log it with `Debug()`, but I got lazy ;) Will probably do this in another patch.